dspace-api/src/main/java/org/dspace/core/HibernateDBConnection.java

```diff
@@ -164,6 +164,83 @@ private void configureDatabaseMode() throws SQLException {
      */
     @Override
     public <E extends ReloadableEntity> void uncacheEntity(E entity) throws SQLException {
-        getSession().evict(entity);        
+        if(entity != null) {
+            if (entity instanceof DSpaceObject) {
+                DSpaceObject dso = (DSpaceObject) entity;
+
+                // The metadatavalue relation has CascadeType.ALL, so they are evicted automatically
+                // and we don' need to uncache the values explicitly.
+
+                if(Hibernate.isInitialized(dso.getHandles())) {
+                    for (Handle handle : Utils.emptyIfNull(dso.getHandles())) {
+                        uncacheEntity(handle);
+                    }
+                }
+
+                //We always need to explicitly check Resource Policies because they could have been fetched by
+                // a direct query and thus not through an initialized collection.
+                for (ResourcePolicy policy : Utils.emptyIfNull(dso.getResourcePolicies())) {
+                    uncacheEntity(policy);
+                }
+            }
+
+            if (entity instanceof Item) {
+                Item item = (Item) entity;
+
+                if(Hibernate.isInitialized(item.getSubmitter())) {
```

Comment:
We are preparing a cache that allows to delete EPersons even if they submitted items. In this case a submitter can be null, so please avoid a NPE here.